lighthouse-core/audits/uses-rel-preconnect.js

```diff
@@ -0,0 +1,143 @@
+   * @return {!boolean}
+   */
+  static hasAlreadyConnectedToOrigin(record) {
+    return (
+      record.timing &&
+      record.timing.dnsEnd - record.timing.dnsStart === 0 &&
+      record.timing.connectEnd - record.timing.connectStart === 0
+    );
+  }
+
+  /**
+   * Check is the connection has started before the socket idle time
+   * @param {!WebInspector.NetworkRequest} record
+   * @param {!WebInspector.NetworkRequest} mainResource
+   * @return {!boolean}
+   */
+  static socketStartTimeIsBelowThreshold(record, mainResource) {
+    return Math.max(0, record._startTime - mainResource._endTime) < PRECONNECT_SOCKET_MAX_IDLE;
+  }
+
+  /**
+   * @param {!Artifacts} artifacts
+   * @return {!AuditResult}
+   */
+  static async audit(artifacts) {
+    const devtoolsLogs = artifacts.devtoolsLogs[UsesRelPreconnectAudit.DEFAULT_PASS];
+    let maxWasted = 0;
+
+    const [networkRecords, mainResource] = await Promise.all([
+      artifacts.requestNetworkRecords(devtoolsLogs),
+      artifacts.requestMainResource(devtoolsLogs),
+    ]);
+
+    const origins = networkRecords
+      .filter(record => {
+        // filter out all resources that have the same origin
+        return (
+          !URL.originsMatch(mainResource.url, record.url) &&
+          // filter out all resources that are loaded by the document
+          record.initiatorRequest() !== mainResource &&
+          // filter out urls that do not have an origin (data, ...)
+          !!URL.getOrigin(record.url) &&
+          // filter out all resources where origins are already resolved
+          !UsesRelPreconnectAudit.hasAlreadyConnectedToOrigin(record) &&
+          // make sure the requests are below the PRECONNECT_SOCKET_MAX_IDLE (15s) mark
+          UsesRelPreconnectAudit.socketStartTimeIsBelowThreshold(record, mainResource)
+        );
+      })
+      .map(record => URL.getOrigin(record.url));
+
+    const preconnectOrigins = new Set(origins);
+    const results = [];
+    preconnectOrigins.forEach(origin => {
+      const records = networkRecords.filter(record => URL.getOrigin(record.url) === origin);
+      if (!records.length) {
+        return;
+      }
+
+      // Sometimes requests are done simultaneous and the connection has not been made
+      // chrome will try to connect for each network record, we get the first record
+      let firstRecordOfOrigin;
+      records.forEach(record => {
+        if (!firstRecordOfOrigin || record._startTime < firstRecordOfOrigin._startTime) {
+          firstRecordOfOrigin = record;
+        }
+      });
+
+      const connectionTime =
+        firstRecordOfOrigin.timing.connectEnd - firstRecordOfOrigin.timing.dnsStart;
+      const timeBetweenMainResourceAndConnectStart =
+        firstRecordOfOrigin._startTime * 1000 -
+        mainResource._endTime * 1000 +
+        firstRecordOfOrigin.timing.connectStart;
+      // calculate delta between connectionTime and timeToConnectionStart from main resource
```

Comment:
this comment is a bit confusing, didn't we already compute the delta and the line below is ensuring we just cap the `connectionTime` savings?